streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -95,7 +96,7 @@
      *          |      | Assigned (3)| <----+
      *          |      +-----+-------+      |
      *          |            |              |
-     *          |            |              |
+     *          |            |--------------+
```

Comment:
What's up with this?

Reply:
I meant to call this out, too. In order to get the task re-initialized after a corruption recovery, we could set the state back to PartitionsAssigned, but I felt that would be confusing in the logs. Instead, I added an extra condition that we'll initialize tasks if there are any that need initialization, even though the thread may already be in running, which means we can have a self-transition from running to running.

Reply:
Where does this self-transition happen exactly? And could/should we detect this case and not call `setState()` for this case instead of allowing the transition?

Reply:
\cc @vvcephei

Reply:
Ah, sorry, I didn't see this follow-up question.

It's here: https://github.com/apache/kafka/blob/568dceb5f61c71428432d23d947c5f9b29fb7bfb/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java#L657

I thought about it, but it seemed to be in violation of the whole idea of having a specified state machine. I.e., saying "`if not running, setState(running)`" is literally the same thing as saying that self-transitions are allowed. So why not just say that self-transitions are allowed (explicitly instead of implicitly)?

Anyway, that's what I was thinking; no guarantee that it makes sense ;)

Reply:
I agree with the sentiment. However, our overall patterns is to disallowed self-transitions, ie, this pattern is applied throughout all other code. IMHO, I might be better to stick with one pattern. Thus, if we think we should allow self-transitions, we might want to do it for all states and update the code throughout accordingly. Begin in a "mixed mode" seems suboptimal.

Of course this would be follow up work. And frankly I am not sure if it would buy us much at this point?

Reply:
FWIW, we actually do follow this pattern already, at least with respect to StreamThread state transitions. I'd rather be consistent with the other StreamThread state changes and inconsistent with other places in the code (eg the Task state changes), especially if we all agree this seems like the sensible approach to the fsm

Reply:
Thanks for chiming in. Seems I did not separate thread state and task state transition. Sorry for the confusion.

Reply:
No worries, thanks for the review @mjsax . Now that I'm thinking about it, it does seem like Task is _allowing_ self-transitions, rather than _disallowing_ them, so we should probably add the self-transitions to the Task state machine. But maybe in a follow-up.